---
title: Homework Assignment 6
author: "Amanda & Delphine"
date: 04/04/2025
format: pdf
header-includes:
   - \usepackage{amsmath}
   - \usepackage{bbm}
   - \usepackage{array}
   - \usepackage{multirow}
   - \usepackage{graphicx}
   - \usepackage{float}
   - \usepackage{apacite}
   - \usepackage{natbib}
execute: 
  echo: true
fontsize: 11pt
geometry: margin = 1in
linestretch: 1.5
---

\newpage
(1)

See below for References.

(2)

In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import matplotlib.cm as cm
import math

from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_samples, silhouette_score
from sklearn.metrics.cluster import rand_score

from sklearn.model_selection import train_test_split, cross_val_score

heart = pd.read_csv("https://raw.githubusercontent.com/PratheepaJ/datasets/refs/heads/master/ass6-dataset.csv")

In [6]:
print(heart.shape)
heart.dtypes

(303, 14)


age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca          float64
thal        float64
num           int64
dtype: object

(1) The classification problem is: classify each patient as having heart disease or no.  

\newpage

(3)

The number of variables is 14, and the number of observations 303. The variable type float is used for oldpeak, Ca and thal.
The target is num, which is the number of heart attacks, stored as int. The integer variable is used for age, sex, cholesterol, fasting blood sugar, resting electrocardiogram, 

(4)

In [16]:
new_heart = heart[(heart["num"] > 0)]
new_heart

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
1     67    1   4       160   286    0        2      108      1      1.5   
2     67    1   4       120   229    0        2      129      1      2.6   
6     62    0   4       140   268    0        2      160      0      3.6   
8     63    1   4       130   254    0        2      147      0      1.4   
9     53    1   4       140   203    1        2      155      1      3.1   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
297   57    0   4       140   241    0        0      123      1      0.2   
298   45    1   1       110   264    0        0      132      0      1.2   
299   68    1   4       144   193    1        0      141      0      3.4   
300   57    1   4       130   131    0        0      115      1      1.2   
301   57    0   2       130   236    0        2      174      0      0.0   

     slope   ca  thal  num  
1        2  3.0   3.0    2  
2        2  2.0   7.0    1  
6        3  2.0   3.0    3  
8        2  1.0   7.0    2  
9        3  0.0   7.0    1  
..     ...  ...   ...  ...  
297      2  0.0   7.0    1  
298      2  0.0   7.0    1  
299      2  2.0   7.0    2  
300      2  1.0   7.0    3  
301      2  1.0   3.0    1  

[139 rows x 14 columns]

0      0
1      2
2      1
3      0
4      0
      ..
298    1
299    2
300    3
301    1
302    0
Name: num, Length: 303, dtype: int64

(5)

More people don't have diabetes than do in the dataset. 

(6)

BMI has 11 missing values, glucose as 5 missing values, blood pressure has 35 missing values, and insulin has 374 missing values. 

(7)

In [ ]:


X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=0, stratify=y)


y_test

10     0
293    2
31     2
149    0
177    1
      ..
162    0
154    3
73     1
282    3
230    0
Name: num, Length: 76, dtype: int64

In [ ]:
y = heart['num']

scalerino = StandardScaler()
X = pd.DataFrame(
    scalerino.fit_transform(heart), 
    index=heart.index, 
    columns=heart.columns
    )

X

age       sex        cp  trestbps      chol       fbs   restecg  \
0    0.948726  0.686202 -2.251775  0.757525 -0.264900  2.394438  1.016684   
1    1.392002  0.686202  0.877985  1.611220  0.760415 -0.417635  1.016684   
2    1.392002  0.686202  0.877985 -0.665300 -0.342283 -0.417635  1.016684   
3   -1.932564  0.686202 -0.165268 -0.096170  0.063974 -0.417635 -0.996749   
4   -1.489288 -1.457296 -1.208521 -0.096170 -0.825922 -0.417635  1.016684   
..        ...       ...       ...       ...       ...       ...       ...   
298 -1.046013  0.686202 -2.251775 -1.234430  0.334813 -0.417635 -0.996749   
299  1.502821  0.686202  0.877985  0.700612 -1.038723  2.394438 -0.996749   
300  0.283813  0.686202  0.877985 -0.096170 -2.238149 -0.417635 -0.996749   
301  0.283813 -1.457296 -1.208521 -0.096170 -0.206864 -0.417635  1.016684   
302 -1.821745  0.686202 -0.165268  0.359134 -1.386944 -0.417635 -0.996749   

      thalach     exang   oldpeak     slope        ca      thal       num  
0    0.017197 -0.696631  1.087338  2.274579 -0.718306  0.653650 -0.764198  
1   -1.821905  1.435481  0.397182  0.649113  2.487269 -0.895552  0.866450  
2   -0.902354  1.435481  1.346147  0.649113  1.418744  1.170051  0.051126  
3    1.637359 -0.696631  2.122573  2.274579 -0.718306 -0.895552 -0.764198  
4    0.980537 -0.696631  0.310912 -0.976352 -0.718306 -0.895552 -0.764198  
..        ...       ...       ...       ...       ...       ...       ...  
298 -0.770990 -0.696631  0.138373  0.649113 -0.718306  1.170051  0.051126  
299 -0.376896 -0.696631  2.036303  0.649113  1.418744  1.170051  0.866450  
300 -1.515388  1.435481  0.138373  0.649113  0.350219  1.170051  1.681773  
301  1.068113 -0.696631 -0.896862  0.649113  0.350219 -0.895552  0.051126  
302  1.024325 -0.696631 -0.896862 -0.976352       NaN -0.895552 -0.764198  

[303 rows x 14 columns]

(8)

(9)

In [30]:
m_log = LogisticRegression(max_iter=120)

(10)

In [38]:
m_log.fit(X_train, y_train)
pred_prob = m_log.predict_proba(X_test)

(11)

In [39]:
m_log_pred = m_log.predict(X_test)
accuracy_score(y_test, m_log_pred) #the default probability cutoff in scikit is 0.5

0.7708333333333334

(12)

(13)

Performance could be improved with more data. 

(14)

(15)

(16)

## References

Diabetes Dataset. (n.d.). Retrieved April 2, 2025, from https://www.kaggle.com/datasets/mathchi/diabetes-data-set
